### ISSUE TITLE
RFE: choosing Service type

### ISSUE THREAD

**Issue by kadel** [kubernetes-incubator/kompose#154]
When I'm using `kompose up` sometimes I need to change Service type.
Right now we generate services that are `ClusterIP`, but sometimes I want to have service exposed to outside of the cluster, than I have to manually change type and set it to `LoadBalancer`.

It would be nice to have a way to specify type of services.

### Proposal:
Use image label in docker-compose.yml to set type of services.
By default we would still create `ClusterIP` services, but when there is a specific label we would create `LoadBalancer`

example:
```yaml
version: "2"
services:
    backend:
        image: backend
        ports:
                - 3000:3000
    frontend:
        image: frontend
        ports:
                - 8080:8080
        environment:
                API_URL: http://backend:3000/api
        labels:
                kompose.service.public: True
```

we could use labels to specify Ingress/Route https://github.com/skippbox/kompose/issues/140

**Comment by dustymabe** [comment]
> Use image label in docker-compose.yml to set type of services.

+1. I wonder if we should make the labels kompose specific though. This would probably be a good thing for @pradeepto to weigh in on.

**Comment by ngtuna** [comment]
Otherwise, we can add `--external` flag ? I am also +1 for the kompose specific label.

**Comment by kadel** [comment]
I was also thinking about about doing this via cmd flag.
But downside of using flag is that you have to always remember to add that flag.

Benefit of doing it via label is that you do it only once. You edit file, commit it and than you can forget about it, and you have everything in one file.

Now when I'm thinking about it,  this could be also in preference file.
Than we can avoid modifying original docker-compose.yml file.

it can be just list of public service names
and for Ingress we can extend it to hostnames
```yaml
expose:
   - foo.bar.example.com=frontend # this will create Ingress/Route
   - frontend # this will create just LoadBalancer service
```

**Comment by surajssd** [comment]
Or how about the user mentions in preference file something like this:

```yaml
profiles:
  default:
    provider: kubernetes
    objects:
      - deployment
      - replicaset
      - ingress
current-context: default
```

**Comment by kadel** [comment]
Ingress can be used with both ClusterIP or LoadBalancer.

**Comment by garethr** [comment]
I hit the same issue today. Labels would work for my usecase. Namespacing them to kompose makes sense.

Note that labels should ideally be reverse DNS namespaced (see https://docs.docker.com/engine/userguide/labels-custom-metadata/), so (unless you also own a specific kompose domain) something like:

```
com.skippbox.kompose.{name of label}
```

**Comment by sebgoa** [comment]
while loadbalancer services are needed in prod, let's not forget about the dev use case who may want to create a nodeport.

**Comment by kadel** [comment]
Agreed, once we do this we should support all three Service types.

**Comment by sebgoa** [comment]
I like the use of labels for this.

**Comment by kadel** [comment]
OK, lets do this via labels.
Using labels to provide additional data for conversion will also solve my problem with #39 :-)

Now we just need to figure out proper namespace for kompose labels as @garethr mentioned ideally it should be reverse DNS

**Comment by procrypt** [comment]
@kadel Can I work on this issue if no one else is working on it.

**Comment by dustymabe** [comment]
Can someone assign this issue to @procrypt ?

**Comment by sebgoa** [comment]
@procrypt I just sent you an invite, you need to accept so that we can assign you tickets.

**Comment by procrypt** [comment]
@sebgoa I have accepted the invite.

**Comment by juanpastas** [comment]
Is not clear for me, can I create an Ingress yet? using `expose`?

**Comment by sebgoa** [comment]
@juanpastas what do you mean by Ingress ?

If you mean an ingress rule (assuming you have an ingress controller running), the answer is no.

Every Docker _service_ in the compose file gets exposed via a Kubernetes service. By default this service is of cluster IP type. If you want to automatically generate a nodeport or load balancer type service you can specify that using a label in your compose file.

**Comment by surajssd** [comment]
@juanpastas you can have ingress created automatically for you once we have PR https://github.com/kubernetes-incubator/kompose/pull/285 merged

**Comment by juanpastas** [comment]
thanks @surajssd.

Do you know how is it possible to:

- Create volumes with an specified size?
- Delete old replica sets when using `kubectl apply -f my.yml`, I read that I need to specify `deployment.spec.revisionHistoryLimit` [here](https://github.com/kubernetes/kubernetes/issues/23597), is this something kompose helps me with?

**Comment by kadel** [comment]
Hi @juanpastas, currently none of those are possible to change with kompose.

In future we might add something to allow specifying size of volumes.
If you want to change those you can use `kompose convert` and then edit files before deploying using `kubectl`
